src/modules/flow/http/http.c

```diff
@@ -0,0 +1,245 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include "sol-flow/http.h"
+#include "sol-flow.h"
+#include "sol-http.h"
+#include "sol-vector.h"
+#include "sol-http-client.h"
+#include "sol-log.h"
+#include "sol-util.h"
+
+struct http_data {
+    struct sol_flow_node *node;
+    char *content_type;
+    char *url;
+    bool strict;
+    struct sol_ptr_vector pending_conns;
+};
+
+struct http_node_type {
+    struct sol_flow_node_type base;
+    void (*response_process_func) (struct sol_http_response *response,
+        struct http_data *mdata);
+};
+
+static void
+http_close(struct sol_flow_node *node, void *data)
+{
+    struct http_data *mdata = data;
+    uint16_t i;
+    struct sol_http_client_connection *conn;
+
+    free(mdata->url);
+    free(mdata->content_type);
+    SOL_PTR_VECTOR_FOREACH_IDX (&mdata->pending_conns, conn, i)
+        sol_http_client_connection_cancel(conn);
+    sol_ptr_vector_clear(&mdata->pending_conns);
+}
+
+static int
+http_open(struct sol_flow_node *node, void *data,
+    const struct sol_flow_node_options *options)
+{
+    struct http_data *mdata = data;
+
+    struct sol_flow_node_type_http_get_string_options *opts =
+        (struct sol_flow_node_type_http_get_string_options *)options;
+
+    if (opts->url) {
+        mdata->url = strdup(opts->url);
+        SOL_NULL_CHECK(mdata->url, -ENOMEM);
+    }
+
+    if (opts->content_type) {
+        mdata->content_type = strdup(opts->content_type);
+        SOL_NULL_CHECK_GOTO(mdata->content_type, err_content_type);
+    }
+
+    mdata->strict = opts->strict;
+    sol_ptr_vector_init(&mdata->pending_conns);
+    mdata->node = node;
+    return 0;
+
+err_content_type:
+    free(mdata->url);
+    return -ENOMEM;
+}
+
+static int
+url_process(struct sol_flow_node *node, void *data, uint16_t port,
+    uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct http_data *mdata = data;
+    const char *url;
+    int r;
+
+    r = sol_flow_packet_get_string(packet, &url);
+    SOL_INT_CHECK(r, < 0, r);
+
+    SOL_DBG("New URL received:%s - old URL:%s", url, mdata->url);
+    free(mdata->url);
+    mdata->url = strdup(url);
+    SOL_NULL_CHECK(mdata->url, -ENOMEM);
+    return 0;
+}
+
+static void
+get_string_process(struct sol_http_response *response, struct http_data *mdata)
+{
+    char *result;
+
+    SOL_DBG("String process");
```

Comment:
make this debug more useful and print the `%.*s` with content.used and content.data